lightly/api/api_workflow_datasources.py

```diff
@@ -51,44 +66,65 @@ def _download_raw_files(
                 cursor=cursor,
                 use_redirected_read_url=use_redirected_read_url,
                 **relevant_filenames_kwargs,
-                **kwargs
+                **kwargs,
             )
             cursor = response.cursor
             samples.extend(response.data)
             if progress_bar is not None:
                 progress_bar.update(len(response.data))
-        samples = [(s.file_name, s.read_url) for s in samples]
+        sample_map = {}
+        for s in samples:
+            if s.file_name.startswith("/"):
+                raise ValueError(
+                        f"Absolute file paths like {s.file_name} are not supported"
+                        f" in relevant filenames file {relevant_filenames_file_name}"
+                )
+            elif s.file_name.startswith(("./", "../")):
+                raise ValueError(
+                        f"Using dot notation ('./', '../') like in {s.file_name} is not supported"
+                        f" in relevant filenames file {relevant_filenames_file_name}"
+                )
+            elif s.file_name in sample_map:
+                warnings.warn(
+                    UserWarning(
+                        f"Duplicate filename {s.file_name} in relevant"
+                        f" filenames file {relevant_filenames_file_name}"
+                    )
+                )
+            else:
+                sample_map[s.file_name] = s.read_url
+        samples = [(file_name, read_url) for file_name, read_url in sample_map.items()]
```

Comment:
please move the logic of this loop in to the above loop after the last statement. No need to double loop over 1000+ entries

Reply:
We cannot move it up, as we need the dictionary in the first loop (for quick lookup of existance), but need to create a list from it later.

Reply:
Python isn't super fast anyway and it's not obvious that you'd get faster execution with a single loop, but you could do something like:

```python
        samples_out=[]
        for s in samples:
            if s.file_name.startswith("/"):
                raise ValueError(
                        f"Absolute file paths like {s.file_name} are not supported"
                        f" in relevant filenames file {relevant_filenames_file_name}"
                )
            elif s.file_name.startswith(("./", "../")):
                raise ValueError(
                        f"Using dot notation ('./', '../') like in {s.file_name} is not supported"
                        f" in relevant filenames file {relevant_filenames_file_name}"
                )
            elif s.file_name in sample_map:
                warnings.warn(
                    UserWarning(
                        f"Duplicate filename {s.file_name} in relevant"
                        f" filenames file {relevant_filenames_file_name}"
                    )
                )
            else:
                sample_map[s.file_name] = s.read_url
                samples_out.append((s.file_name, s.read_url))
        return samples_out
```

The thing is, Python lists abstract away lots of implementation detail, and may end up getting copied in memory multiple times with all those append operations. It really isn't obviously faster to iterate once. We could make this change though.

Reply:
The fastest implementation (though I doubt this is a performance critical code snippet) should be:
```python
        samples_out = [None]*len(samples)
        for s in samples:
            if s.file_name.startswith("/"):
                raise ValueError(
                        f"Absolute file paths like {s.file_name} are not supported"
                        f" in relevant filenames file {relevant_filenames_file_name}"
                )
            elif s.file_name.startswith(("./", "../")):
                raise ValueError(
                        f"Using dot notation ('./', '../') like in {s.file_name} is not supported"
                        f" in relevant filenames file {relevant_filenames_file_name}"
                )
            elif s.file_name in sample_map:
                warnings.warn(
                    UserWarning(
                        f"Duplicate filename {s.file_name} in relevant"
                        f" filenames file {relevant_filenames_file_name}"
                    )
                )
            else:
                sample_map[s.file_name] = s.read_url
                samples_out.append((s.file_name, s.read_url))
        return samples_out
```

Reply:
Implementation updated for looping once. It should be a little faster, though again, I don't believe this is performance critical (if it was we probably wouldn't want to use Python for it). Do you think this is comparably readable & potentially worth it @MalteEbner @japrescott ?

Reply:
This is not performance critical. I would keep the current method. Without actually measuring it, we don't know which method is faster. E.g. we could be just trading number of computations against cache coherence.